### ISSUE TITLE
Tier 3 OpenBSD Support for x86_64

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#2016]
Here's the issue to track OpenBSD support. The current goal is Tier 3, since Zig can't even build on OpenBSD yet. Once Zig reaches Tier 3 this issue will get renamed to Tier 2, and so on. @Aversiste has started the effort in #1921.

**Comment by andrewrk** [comment]
Here is some partially finished code that one can use for inspiration: #1921

**Comment by andrewrk** [comment]
SourceHut has OpenBSD images now: https://man.sr.ht/builds.sr.ht/compatibility.md#openbsd

**Comment by voutilad** [comment]
FWIW, OpenBSD recently imported LLVM 8.0.0 into base: https://marc.info/?l=openbsd-cvs&m=156132754603489&w=2

It's available now in the latest -current snapshots for anyone interested in this issue.

**Comment by justjosias** [comment]
I am willing to help out with OpenBSD support. I have looked through the previous PRs (particularly #5324), and would like to tackle it, getting Zig to even compile on OpenBSD and getting those changes merged into master.

**Comment by andrewrk** [comment]
Happy to hear that @justjosias . My suggestion is to make smaller PRs that can be merged independently, making incremental progress. I think you will meet with swift success if you do this.

**Comment by kivikakk** [comment]
@justjosias :wave: also keen to see Zig on OpenBSD; happy to work with you on testing/iterating on your changes.

**Comment by gaultier** [comment]
OpenBSD-current just added llvm 10 to their (binary) packages. Maybe this could rejuvenile this effort?

**Comment by justjosias** [comment]
@gaultier You have to compile LLVM from source, which isn't so hard (the boostrap works very well, just need to add some linker and header include paths). At that point you have to patch and modify Zig to get it to build, which is what this issue is about.

**Comment by gaultier** [comment]
@justjosias Right, got it, I did just this and I realized OpenBSD is not in the list of supported OSes in the zig source code.

**Comment by semarie** [comment]
FYI I started a port of zig for OpenBSD : https://github.com/semarie/zig/tree/openbsd
it is based on LLVM version from ports (full llvm-10), and I am able to build a stage2 zig compiler.

some work is still need:
- linkage (to build zig compiler I had to explicitly pass c++ libraries)
- proper `selfExePath` implementation (OpenBSD doesn't provide such facility at all)
- linkage errors will running test-behavior
- etc...

**Comment by jedisct1** [comment]
Awesome news, thank you @semarie!

**Comment by vazub** [comment]
@semarie are you still working on this? looks like your repo is not there anymore

**Comment by semarie** [comment]
I build zig semi-regulary on OpenBSD using upstream repository without issues (and I sent PR to fix them when I saw a problem).

and OpenBSD has a port for zig in official OpenBSD port-tree (but it is disabled due to too big memory requirement to build it from scratch).

**Comment by vazub** [comment]
@semarie Does this mean this ticket can be closed as Done now? Or is there anything pending from the list you mentioned above? I assume selfExePath resolution might still be the issue?

**Comment by semarie** [comment]
zig std has an alternative code for `selfExePath` as best-effort.

my last build was one month ago (with repository at 99c3578f697fe0b0151049f74208b86244b4d171). for me, it should be marked as Done, even if currently is doesn't build (see #13565 where OpenBSD is also impacted).

**Comment by vazub** [comment]
Looks like the fixes for the blockers have been merged @semarie

**Comment by andrewrk** [comment]
> but it is disabled due to too big memory requirement to build it from scratch

In case you [missed it](https://ziglang.org/news/goodbye-cpp/), #6485 is resolved now :-)

**Comment by kuon** [comment]
I sadly do not have a lot of time to contribute code to this issue, but I'd like to let you know that we use OpenBSD quite a bit for servers and routers, and I'll be happy to help testing.

**Comment by LowByteFox** [comment]
I am currently running OpenBSD 7.3-Current and I've tried to compile zig 0.11.0. It's not possible to compile `getFdPath`. Before I tried to compile zig @oven-sh uses because of the stage 3 compiler segfault-ing.

What got me confusing is that on release notes of zig 0.11.0 under Tier 2, OpenBSD has ✅ but should have 📖 because it doesn't compile. I don't know if got there by accident or you just cross-compiled it without being on OpenBSD, I don't know how you got it.

If someone wants to compile it themselves, I provide llvm16 pre-built binary for [amd64 OpenBSD 7.3-current](https://github.com/Bun-OpenBSD/llvm) in releases. I compiled it as zig's wiki tells me. + Don't forget to install it into /usr/local/llvm16

Before compiling zig, edit `/etc/login.conf` and for `staff` increase `datasize-cur` to `6000M` just so clang doesn't fail with Out of Memory error.

My configure command was
```bash
cmake .. -DZIG_STATIC_LLVM=ON -DZIG_STATIC_ZSTD=ON -DZIG_STATIC_ZLIB=ON -DZIG_USE_LLVM_CONFIG=OFF  -DCMAKE_PREFIX_PATH="/usr/local/llvm16;/usr/local"
```
Then run `make edit_cache` and enter advanced mode
under `CMAKE_EXE_LINKER_FLAGS` add `-lexecinfo` so when zig 2 will build it doesn't fail with missing `backtrace`

Now you are good to go, the compilation should fail with
![image](https://github.com/ziglang/zig/assets/64907994/c24c504e-6229-408a-8914-d8ad8ba3ffe7)

**Comment by LowByteFox** [comment]
I am back with more info, okay so I tried to cross compile simple hello world application. But it crashes with `SIGSEGV`

I ssh mounted a computer running OpenBSD and modified output of `zig libc` just so it matches ssh root mount. Then I simply executed
`zig build --libc openbsdlibc.txt -Dtarget=x86_64-openbsd -Doptimize=ReleaseSmall`

Reason why `ReleaseSmall` is because it doesn't require the missing `getcontext` symbol and compiles. Then I just copied the output executable to the computer and executed it. It just segfaulted.

It could either be because of the "nasty fix" or who knows what.
I'll share my ktrace output [here](https://gist.github.com/LowByteFox/e87755e9cb6eaf7fb2c749bf2b2eb191)

**Comment by bilaliscarioth** [comment]
Found a little issue when `std:Build.LibExeObjStep.linkSystemLibrary` is used on OpenBSD with zig (master branch)

idk for which reason that doesn't work well because every function used on Zig come from of a C Static library ( example SDL2/GL/X11...) are `undefined symbols`

Actually works if i used  `std:Build.LibExeObjStep.addObjectFile` where i put the paths of the object (dynamic or not)
